src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -321,138 +384,104 @@ int MPL_gavl_tree_free(MPL_gavl_tree_t gavl_tree)
                 else
                     cur_node->right = NULL;
             }
-            if (gavl_tree_iptr->gavl_free_fn)
-                gavl_tree_iptr->gavl_free_fn((void *) dnode->val);
+            if (tree_ptr->gavl_free_fn)
+                tree_ptr->gavl_free_fn((void *) dnode->val);
             MPL_free(dnode);
         }
     }
-    MPL_free(gavl_tree_iptr);
+    MPL_free(tree_ptr);
     return mpl_err;
 }
 
-int MPL_gavl_tree_delete(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len)
+static void gavl_tree_delete_node(gavl_tree_s * tree_ptr, gavl_tree_node_s * dnode, void **val)
 {
-    int mpl_err = MPL_SUCCESS;
-    gavl_tree_node_s *cur_node;
     gavl_tree_node_s *inorder_node;
-    gavl_tree_s *gavl_tree_iptr = (gavl_tree_s *) gavl_tree;
 
-    DECLARE_STACK(gavl_tree_node_s *, node_stack);
+    if (dnode->right == NULL) {
+        /* no right child, next inorder node is parent */
+        if (dnode->parent == NULL) {
+            /* delete root node */
+            if (dnode->left) {
+                tree_ptr->root = dnode->left;
+                tree_ptr->root->parent = NULL;
+            } else {
+                tree_ptr->root = NULL;
+            }
+            *val = (void *) dnode->val;
+            MPL_free(dnode);
+        } else {
```

Comment:
Setting *val and free(dnode) seem to be a common path for all branches. Can you move it after the if branchs?

Reply:
I move `MPL_free(dnode)` out, but setting `*val` should not be moved out because we need to replace `dnode` with `inorder_node` when `dnode->right != NULL`, `val` should be set before `dnode` is replaced.

Reply:
that is a standard operation during AVL tree deletion. If we want to delete a node and this node has right child, we should find next inorder node to replace it (or say overwrite it). If you free the deleted node directly, our traverse trace will be broken and rebalance routine will be incorrect in that way.

Reply:
"Standard operation during AVL deletion", "traverse trace will be broken and rebalance routine will be incorrect in that way" do not explain why we need replace `inorder_node ` with `dnode` rather than deleting it.

Reply:
Could you please clarify 
your question is replace `inorder_node` with `dnode` 
or replace `dnode` with `inorder_node`?

Reply:
Oh, I see what you are doing here now. You move the buffer object in the next `inorder_node` into the location of `dnode`, then free the tree node of `inorder_node`. The original buffer object in `dnode` is returned and deleted. 

This comment is confusing to me: "/* find next inorder node and replace deleted node with it */"

Besides, "/* inorder node's parent is dnode, just set dnode right child as NULL */" is redundant as you already explained on top of `if (inorder_node->parent != dnode) {` .

You might also want to change `if (inorder_node->parent != dnode) {` to `if (inorder_node->parent == dnode) {` and swap the two branches to match the sequence of comment.

Reply:
sorry for this confusion. I should make it more concrete.

Reply:
The following comment is not addressed.
>You might also want to change if (inorder_node->parent != dnode) { to if (inorder_node->parent == dnode) { and swap the two branches to match the sequence of comment.

Besides, I think the comments can be also compressed. How about:
1.  Compress
```
        /* first find next inorder node; then we free buffer object of deleted node,
         * copy the content of inorder node into deleted node; finally free inorder
         * node. */
```
to
```
   /* find the next inorder node and move its buffer objects to dnode.
       The original buffer object in dnode is freed. */
```

2. Compress
```
        /* we find the inorder node and check whether inorder's parent is dnode; if not,
         * we need to update inorder_node's parent left child as dnode's right child; if
         * it is, we need to set dnode right branch as NULL; */
```
to
```
  /* remove inorder_node from the tree. */
```

3. Then add comment:
```
        val = (void *) dnode->val;
        /* move inorder_node's buffer object to dnode and then free inorder_node */
        dnode->addr = inorder_node->addr;
        dnode->len = inorder_node->len;
        dnode->val = inorder_node->val;
        dnode = inorder_node;
```